README.md
# PHP HTTP Status Code library

[![Latest Stable Version](https://poser.pugx.org/josantonius/http-status-code/v/stable)](https://packagist.org/packages/josantonius/http-status-code)
[![License](https://poser.pugx.org/josantonius/http-status-code/license)](LICENSE)
[![Total Downloads](https://poser.pugx.org/josantonius/http-status-code/downloads)](https://packagist.org/packages/josantonius/http-status-code)
[![CI](https://github.com/josantonius/php-http-status-code/actions/workflows/ci.yml/badge.svg?branch=main)](https://github.com/josantonius/php-http-status-code/actions/workflows/ci.yml)
[![CodeCov](https://codecov.io/gh/josantonius/php-http-status-code/branch/main/graph/badge.svg)](https://codecov.io/gh/josantonius/php-http-status-code)
[![PSR1](https://img.shields.io/badge/PSR-1-f57046.svg)](https://www.php-fig.org/psr/psr-1/)
[![PSR4](https://img.shields.io/badge/PSR-4-9b59b6.svg)](https://www.php-fig.org/psr/psr-4/)
[![PSR12](https://img.shields.io/badge/PSR-12-1abc9c.svg)](https://www.php-fig.org/psr/psr-12/)

**Translations**: [Español](.github/lang/es-ES/README.md)

PHP library to get HTTP status code messages and definitions.

---

- [Requirements](#requirements)
- [Installation](#installation)
- [Available Classes](#available-classes)
  - [HttpStatusCode Class](#httpstatuscode-class)
- [Exceptions Used](#exceptions-used)
- [Usage](#usage)
- [List in JSON format](#list-in-json-format)
- [Tests](#tests)
- [TODO](#todo)
- [Changelog](#changelog)
- [Contribution](#contribution)
- [Sponsor](#sponsor)
- [License](#license)

---

## Requirements

- Operating System: Linux | Windows.

- PHP versions: 8.0 | 8.1 | 8.2.

## Installation

The preferred way to install this extension is through [Composer](http://getcomposer.org/download/).

To install **PHP HTTP Status Code library**, simply:

```console
composer require josantonius/http-status-code
```

The previous command will only install the necessary files,
if you prefer to **download the entire source code** you can use:

```console
composer require josantonius/http-status-code --prefer-source
```

You can also **clone the complete repository** with Git:

```console
git clone https://github.com/josantonius/http-status-code.git
```

## Available Classes

### HttpStatusCode Class

`Josantonius\HttpStatusCode\HttpStatusCode`

Create a new instance defining the language:

```php
/**
 * @param string $language The language in which the data will be retrieved.
 *
 *                         Available languages: en (English)
 *                                              es (Spanish)
 * 
 * @throws UnsupportedLanguageException if language is not supported.
 */
public function __construct(string $language = 'en');
```

Gets message of an HTTP status code:

```php
public function getMessage(int $code): string|null;
```

Gets the messages of all HTTP status codes:

```php
public function getMessages(): array;
```

Gets definition of an HTTP status code:

```php
public function getDefinition(int $code): string|null;
```

Gets the definitions of all HTTP status codes:

```php
public function getDefinitions(): array;
```

Gets messages and definitions of all HTTP status codes:

```php
public function getAll(): array;
```

## Exceptions Used

```php
use Josantonius\HttpStatusCode\UnsupportedLanguageException;
```

## Usage

Example of use for this library:

### Create a new instance defining the language

```php
use Josantonius\HttpStatusCode\HttpStatusCode;

$httpStatusCode = new HttpStatusCode();     // Gets the messages in English.
```

```php
use Josantonius\HttpStatusCode\HttpStatusCode;

$httpStatusCode = new HttpStatusCode('es'); // Gets the messages in Spanish.
```

### Gets message of an HTTP status code

```php
use Josantonius\HttpStatusCode\HttpStatusCode;

$httpStatusCode = new HttpStatusCode();

$httpStatusCode->getMessage(404);
```

Returns:

```php
'Not Found'
```

### Gets the messages of all HTTP status codes

```php
use Josantonius\HttpStatusCode\HttpStatusCode;

$httpStatusCode = new HttpStatusCode();

$httpStatusCode->getMessages();
```

Returns:

```php
[
    100 => 'Continue',
    101 => 'Switching Protocols',
    102 => 'Processing',

    (...)
]
```

### Gets definition of an HTTP status code

```php
use Josantonius\HttpStatusCode\HttpStatusCode;

$httpStatusCode = new HttpStatusCode();

$httpStatusCode->getDefinition(404);
```

Returns:

```php
'The requested resource could not be found but (...)'
```

### Gets the definitions of all HTTP status codes

```php
use Josantonius\HttpStatusCode\HttpStatusCode;

$httpStatusCode = new HttpStatusCode();

$httpStatusCode->getDefinitions();
```

Returns:

```php
[
    100 => 'The server has received the request (...)',
    101 => 'The requester has asked the server (...)',
    102 => 'A WebDAV request may contain many (...)',

    (...)
]
```

### Gets messages and definitions of all HTTP status codes

```php
use Josantonius\HttpStatusCode\HttpStatusCode;

$httpStatusCode = new HttpStatusCode();

$httpStatusCode->getAll();
```

```php
[
    100 => [
        'message'    => 'Continue',
        'definition' => 'The server has received the request (...)',
    ],
    101 => [
        'message'    => 'Switching Protocols',
        'definition' => 'The requester has asked the server (...)',
    ],
    102 => [
        'message'    => 'Processing',
        'definition' => 'A WebDAV request may contain many (...)',
    ],

    (...)
]
```

## List in JSON format

<https://gist.github.com/Josantonius/0a889ab6f18db2fcefda15a039613293>

## Tests

To run [tests](tests) you just need [composer](http://getcomposer.org/download/)
and to execute the following:

```console
git clone https://github.com/josantonius/php-http-status-code.git
```

```console
cd php-http-status-code
```

```console
composer install
```

Run unit tests with [PHPUnit](https://phpunit.de/):

```console
composer phpunit
```

Run code standard tests with [PHPCS](https://github.com/squizlabs/PHP_CodeSniffer):

```console
composer phpcs
```

Run [PHP Mess Detector](https://phpmd.org/) tests to detect inconsistencies in code style:

```console
composer phpmd
```

Run all previous tests:

```console
composer tests
```

## TODO

- [ ] Add new feature
- [ ] Improve tests
- [ ] Improve documentation
- [ ] Improve English translation in the README file
- [ ] Refactor code for disabled code style rules (see phpmd.xml and phpcs.xml)

## Changelog

Detailed changes for each release are documented in the
[release notes](https://github.com/josantonius/php-http-status-code/releases).

## Contribution

Please make sure to read the [Contributing Guide](.github/CONTRIBUTING.md), before making a pull
request, start a discussion or report a issue.

Thanks to all [contributors](https://github.com/josantonius/php-http-status-code/graphs/contributors)! :heart:

## Sponsor

If this project helps you to reduce your development time,
[you can sponsor me](https://github.com/josantonius#sponsor) to support my open source work :blush:

## License

This repository is licensed under the [MIT License](LICENSE).

Copyright © 2016-present, [Josantonius](https://github.com/josantonius#contact)


CHANGELOG.md
# CHANGELOG

## [v2.0.4](https://github.com/josantonius/php-http-status-code/releases/tag/v2.0.4) (2022-09-29)

* The notation type in the test function names has been changed from camel to snake case for readability.

* Functions were added to document the methods and avoid confusion.

* Disabled the ´CamelCaseMethodName´ rule in ´phpmd.xml´ to avoid warnings about function names in tests.

* The alignment of the asterisks in the comments has been fixed.

* Tests for Windows have been added.

* Tests for PHP 8.2 have been added.

## [v2.0.3](https://github.com/josantonius/php-http-status-code/releases/tag/v2.0.3) (2022-08-11)

* Improved documentation.

## [v2.0.2](https://github.com/josantonius/php-http-status-code/releases/tag/v2.0.2) (2022-08-04)

* Fix Packagist URL in `README.md` files

## [v2.0.1](https://github.com/josantonius/php-http-status-code/releases/tag/v2.0.1) (2022-07-31)

* The links in the `CHANGELOG.md` file have been fixed.

* Removed the warning about version 1.x in the `README.md` files.

* The comment in the exception class was modified.

* pre-release/v2.0.1 in <https://github.com/josantonius/php-http-status-code/pull/2>

**Full Changelog**: <https://github.com/josantonius/php-http-status-code/compare/v2.0.0...v2.0.1>

## [v2.0.0](https://github.com/josantonius/php-http-status-code/releases/tag/v2.0.0) (2022-07-31)

> Version 1.x is considered as deprecated and unsupported.
> In this version (2.x) the library was completely restructured.
> It is recommended to review the documentation for this version and make the necessary changes
> before starting to use it, as it not be compatible with version 1.x.

---

* The library was completely refactored.

* Support for PHP version 8.0.

* Support for earlier versions of PHP **8.0** is discontinued.

* Replaced all static methods in `Josantonius\HttpStatusCode\HttpStatusCode` class.

* Improved documentation; `README.md`, `CODE_OF_CONDUCT.md`, `CONTRIBUTING.md` and `CHANGELOG.md`.

* Removed `Codacy`.

* Removed `PHP Coding Standards Fixer`.

* The `master` branch was renamed to `main`.

* The `develop` branch was added to use a workflow based on `Git Flow`.

* `Travis` is discontinued for continuous integration. `GitHub Actions` will be used from now on.

* Added `.github/CODE_OF_CONDUCT.md` file.
* Added `.github/CONTRIBUTING.md` file.
* Added `.github/FUNDING.yml` file.
* Added `.github/workflows/ci.yml` file.
* Added `.github/lang/es-ES/CODE_OF_CONDUCT.md` file.
* Added `.github/lang/es-ES/CONTRIBUTING.md` file.
* Added `.github/lang/es-ES/LICENSE` file.
* Added `.github/lang/es-ES/README` file.

* Deleted `.travis.yml` file.
* Deleted `.editorconfig` file.
* Deleted `CONDUCT.MD` file.
* Deleted `README-ES.MD` file.
* Deleted `.php_cs.dist` file.

## [1.1.6](https://github.com/josantonius/php-http-status-code/releases/tag/1.1.6) (2018-01-06)

* The tests were fixed.

* Changes in documentation.

## [1.1.5](https://github.com/josantonius/php-http-status-code/releases/tag/1.1.5) (2017-11-08)

* Implemented `PHP Mess Detector` to detect inconsistencies in code styles.

* Implemented `PHP Code Beautifier and Fixer` to fixing errors automatically.

* Implemented `PHP Coding Standards Fixer` to organize PHP code automatically according to PSR standards.

## [1.1.4](https://github.com/josantonius/php-http-status-code/releases/tag/1.1.4) (2017-10-30)

* Implemented `PSR-4 autoloader standard` from all library files.

* Implemented `PSR-2 coding standard` from all library PHP files.

* Implemented `PHPCS` to ensure that PHP code complies with `PSR2` code standards.

* Implemented `Codacy` to automates code reviews and monitors code quality over time.

* Implemented `Codecov` to coverage reports.

* Added `HTTPStatusCode/phpcs.ruleset.xml` file.

* Deleted `HTTPStatusCode/src/bootstrap.php` file.

* Deleted `HTTPStatusCode/tests/bootstrap.php` file.

* Deleted `HTTPStatusCode/tests/resources/http-status-code.jsond` file.

* Deleted `HTTPStatusCode/vendor` folder.

* Changed `Josantonius\HTTPStatusCode\Test\HTTPStatusCodeTest` class to  `Josantonius\HTTPStatusCode\HTTPStatusCodeTest` class.

* Added `Josantonius\HTTPStatusCode\HTTPStatusCodeCollection` class.
* Added `Josantonius\HTTPStatusCode\HTTPStatusCodeCollection::all()` method.
* Added `Josantonius\HTTPStatusCode\HTTPStatusCodeCollection::get()` method.

* Deleted `Josantonius\HTTPStatusCode\HTTPStatusCode::load()` method.

## [1.1.3](https://github.com/josantonius/php-http-status-code/releases/tag/1.1.3) (2017-09-13)

* Unit tests supported by `PHPUnit` were added.

* The repository was synchronized with Travis CI to implement continuous integration.

* Added `HTTPStatusCode/src/bootstrap.php` file

* Added `HTTPStatusCode/tests/bootstrap.php` file.

* Added `HTTPStatusCode/phpunit.xml.dist` file.
* Added `HTTPStatusCode/_config.yml` file.
* Added `HTTPStatusCode/.travis.yml` file.

* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest` class.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testLoad()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testLoadES()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetEN()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetES()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetLargeEN()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetLargeES()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetShortEN()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetShortES()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetUndefinedEN()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetUndefinedES()` method.
* Deleted `Josantonius\HTTPStatusCode\Tests\HTTPStatusCodeTest::testGetAllEN()` method.

src/data/messages/en.php
<?php

/*
* This file is part of https://github.com/josantonius/php-http-status-code repository.
*
* (c) Josantonius <hello@josantonius.dev>
*
* For the full copyright and license information, please view the LICENSE
* file that was distributed with this source code.
*
* @see https://en.wikipedia.org/wiki/List_of_HTTP_status_codes
*/

return [
    100 => 'Continue',
    101 => 'Switching Protocols',
    102 => 'Processing',
    103 => 'Checkpoint',
    200 => 'OK',
    201 => 'Created',
    202 => 'Accepted',
    203 => 'Non-Authoritative Information',
    204 => 'No Content',
    205 => 'Reset Content',
    206 => 'Partial Content',
    207 => 'Multi-Status',
    208 => 'Already Reported',
    300 => 'Multiple Choices',
    301 => 'Moved Permanently',
    302 => 'Found',
    303 => 'See Other',
    304 => 'Not Modified',
    305 => 'Use Proxy',
    306 => 'Switch Proxy',
    307 => 'Temporary Redirect',
    308 => 'Permanent Redirect',
    400 => 'Bad Request',
    401 => 'Unauthorized',
    402 => 'Payment Required',
    403 => 'Forbidden',
    404 => 'Not Found',
    405 => 'Method Not Allowed',
    406 => 'Not Acceptable',
    407 => 'Proxy Authentication Required',
    408 => 'Request Time-out',
    409 => 'Conflict',
    410 => 'Gone',
    411 => 'Length Required',
    412 => 'Precondition Failed',
    413 => 'Request Entity Too Large',
    414 => 'Request-URI Too Long',
    415 => 'Unsupported Media Type',
    416 => 'Requested Range Not Satisfiable',
    417 => 'Expectation Failed',
    418 => "I'm a teapot",
    421 => 'Unprocessable Entity',
    422 => 'Misdirected Request',
    423 => 'Locked',
    424 => 'Failed Dependency',
    426 => 'Upgrade Required',
    428 => 'Precondition Required',
    429 => 'Too Many Requests',
    431 => 'Request Header Fields Too Large',
    451 => 'Unavailable For Legal Reasons',
    500 => 'Internal Server Error',
    501 => 'Not Implemented',
    502 => 'Bad Gateway',
    503 => 'Service Unavailable',
    504 => 'Gateway Timeout',
    505 => 'HTTP Version Not Supported',
    506 => 'Variant Also Negotiates',
    507 => 'Insufficient Storage',
    508 => 'Loop Detected',
    509 => 'Bandwidth Limit Exceeded',
    510 => 'Not Extended',
    511 => 'Network Authentication Required',
];
